src/nbl/video/CVulkanImage.cpp

```diff
@@ -2,41 +2,43 @@
 
-namespace nbl {
+namespace nbl
+{
 namespace video
 {
-    CVulkanImage::~CVulkanImage()
-    {
-        auto* vk = m_vkdevice->getFunctionTable();
-        auto vkdev = m_vkdevice->getInternalObject();
-
-        vk->vk.vkDestroyImage(vkdev, m_vkimg, nullptr);
-    }
-
-    CVulkanImage::CVulkanImage(CVKLogicalDevice* _vkdev, IGPUImage::SCreationParams&& _params) : IGPUImage(std::move(_params)), m_vkdevice(_vkdev)
-    {
-        auto* vk = m_vkdevice->getFunctionTable();
-        auto vkdev = m_vkdevice->getInternalObject();
-
-        VkImageCreateInfo ci;
-        ci.pNext = nullptr;
-        ci.sType = VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO;
-        ci.arrayLayers = params.arrayLayers;
-        ci.extent = { params.extent.width, params.extent.height, params.extent.depth };
-        ci.flags = static_cast<VkImageCreateFlags>(params.flags);
-        ci.format = static_cast<VkFormat>(params.format);
-        ci.imageType = static_cast<VkImageType>(params.type);
-        ci.initialLayout = static_cast<VkImageLayout>(params.initialLayout);
-        ci.mipLevels = params.mipLevels;
-        ci.samples = static_cast<VkSampleCountFlagBits>(params.samples);
-        ci.sharingMode = VK_SHARING_MODE_EXCLUSIVE;
-        ci.pQueueFamilyIndices = params.queueFamilyIndices->data();
-        ci.queueFamilyIndexCount = params.queueFamilyIndices->size();
-        ci.tiling = static_cast<VkImageTiling>(params.tiling);
-        ci.usage = static_cast<VkImageUsageFlags>(params.usage);
-
-        vk->vk.vkCreateImage(vkdev, &ci, nullptr, &m_vkimg);
-    }
+
+CVulkanImage::~CVulkanImage()
+{
+    auto* vk = m_vkdevice->getFunctionTable();
+    auto vkdev = m_vkdevice->getInternalObject();
+
+    vk->vk.vkDestroyImage(vkdev, m_vkimg, nullptr);
+}
+
+CVulkanImage::CVulkanImage(CVKLogicalDevice* _vkdev, IGPUImage::SCreationParams&& _params) : IGPUImage(_vkdev, std::move(_params)), m_vkdevice(_vkdev)
+{
+    auto* vk = m_vkdevice->getFunctionTable();
+    auto vkdev = m_vkdevice->getInternalObject();
+
+    VkImageCreateInfo ci;
+    ci.pNext = nullptr;
+    ci.sType = VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO;
+    ci.arrayLayers = params.arrayLayers;
+    ci.extent = { params.extent.width, params.extent.height, params.extent.depth };
+    ci.flags = static_cast<VkImageCreateFlags>(params.flags);
+    ci.format = static_cast<VkFormat>(params.format);
+    ci.imageType = static_cast<VkImageType>(params.type);
+    ci.initialLayout = static_cast<VkImageLayout>(params.initialLayout);
+    ci.mipLevels = params.mipLevels;
+    ci.samples = static_cast<VkSampleCountFlagBits>(params.samples);
+    ci.sharingMode = VK_SHARING_MODE_EXCLUSIVE;
+    ci.pQueueFamilyIndices = params.queueFamilyIndices->data();
+    ci.queueFamilyIndexCount = params.queueFamilyIndices->size();
+    ci.tiling = static_cast<VkImageTiling>(params.tiling);
+    ci.usage = static_cast<VkImageUsageFlags>(params.usage);
+
+    vk->vk.vkCreateImage(vkdev, &ci, nullptr, &m_vkimg);
```

Comment:
you want to fire this off inside `CVulkanLogicalDevice` not inside the CVulkanImage

if the function doesn't fail, then pass all the things you need to the constructor (which is noexcept)

Reply:
Do you want this in `CVKLogicalDevice::createGPUImageOnDedMem`, it also takes a `IDriverMemoryBacked::SDriverMemoryRequirements` param, or should I make a `CVKLogicalDevice::createImage`?

Reply:
Vulkan impl shall actually implement all the `create` methods, allowing for creation separated from memory binding

`create*OnDedMem` should call both vanilla `create` and then bindMemory (with dedicated memory) instead of what OpenGL does, please google "Vulkan Dedicated Memory Allocation".